connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/ProtobufSerializer.scala

```diff
@@ -0,0 +1,250 @@
+          Timestamp.newBuilder()
+            .setSeconds((millis / 1000))
+            .setNanos(((millis % 1000) * 1000000).toInt)
+            .build()
+
+      case (ArrayType(et, containsNull), _) =>
+        val elementConverter =
+          newConverter(et, fieldDescriptor, catalystPath :+ "element", protoPath :+ "element")
+        (getter, ordinal) => {
+          val arrayData = getter.getArray(ordinal)
+          val len = arrayData.numElements()
+          val result = new Array[Any](len)
+          var i = 0
+          while (i < len) {
+            if (containsNull && arrayData.isNullAt(i)) {
+              result(i) = null
+            } else {
+              result(i) = elementConverter(arrayData, i)
+            }
+            i += 1
+          }
+          // Protobuf writer is expecting a Java Collection, so we convert it into
+          // `ArrayList` backed by the specified array without data copying.
+          java.util.Arrays.asList(result: _*)
+        }
+
+      case (st: StructType, MESSAGE) =>
+        val structConverter =
+          newStructConverter(st, fieldDescriptor.getMessageType, catalystPath, protoPath)
+        val numFields = st.length
+        (getter, ordinal) => structConverter(getter.getStruct(ordinal, numFields))
+
+      case (MapType(kt, vt, valueContainsNull), MESSAGE) =>
+        var keyField: FieldDescriptor = null
+        var valueField: FieldDescriptor = null
+        fieldDescriptor.getMessageType.getFields.asScala.map { field =>
+          field.getName match {
+            case "key" =>
+              keyField = field
+            case "value" =>
+              valueField = field
+          }
+        }
+
+        val keyConverter = newConverter(kt, keyField, catalystPath :+ "key", protoPath :+ "key")
+        val valueConverter =
+          newConverter(vt, valueField, catalystPath :+ "value", protoPath :+ "value")
+
+        (getter, ordinal) =>
+          val mapData = getter.getMap(ordinal)
+          val len = mapData.numElements()
+          val list = new java.util.ArrayList[DynamicMessage]()
+          val keyArray = mapData.keyArray()
+          val valueArray = mapData.valueArray()
+          var i = 0
+          while (i < len) {
+            val result = DynamicMessage.newBuilder(fieldDescriptor.getMessageType)
+            if (valueContainsNull && valueArray.isNullAt(i)) {
+              result.setField(keyField, keyConverter(keyArray, i))
+              result.setField(valueField, valueField.getDefaultValue)
+            } else {
+              result.setField(keyField, keyConverter(keyArray, i))
+              result.setField(valueField, valueConverter(valueArray, i))
+            }
+            list.add(result.build())
+            i += 1
+          }
+          list
+      case (_: YearMonthIntervalType, INT) =>
+        (getter, ordinal) => getter.getInt(ordinal)
+
+      case (_: DayTimeIntervalType, LONG) =>
```

Comment:
We can map it to duration: https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#duration